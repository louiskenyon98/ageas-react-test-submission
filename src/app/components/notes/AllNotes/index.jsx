import React from 'react';
import {Link} from 'react-router-dom';
import styles from '../../../../themes/style.scss';

class AllNotes extends React.Component {
    renderAll() {
        //Map array in reverse order so as to render the newest notes first.
        //Shallow copy so as not to alter original array.
        return this.props.notes.slice(0).reverse().map((note) => {
            return (
                <div className={`col-xs-1 col-md-4 col-xl-3 d-flex align-items-stretch ${styles.column}`} key={note.id}>
                    <div className={`ui raised card ${styles.card}`}>
                        <div className="content">
                            <i
                                onClick={() => this.props.deleteNote(note.id)}
                                className={`right floated close icon ${styles["delete-button"]}`}
                            />
                            <div className={`header ${styles.header}`}>
                                {note.title}
                            </div>
                            <div className={`description ${styles.body}`}>
                                <p>{note.body}</p>
                            </div>
                        </div>
                        <div className="extra content">
                            <span className="right floated pencil alternate">
                                <Link to={`/edit/${note.id}`}>
                                    <i className={`pencil alternate icon ${styles["edit-button"]}`}/>
                                </Link>
                            </span>
                        </div>
                    </div>
                </div>
            )
        })
    }

    render() {
        return (
            <div className={`container-fluid`}>
                <div className={`row`}>
                    {this.renderAll()}
                </div>
            </div>
        )
    }
}

export default AllNotes;
