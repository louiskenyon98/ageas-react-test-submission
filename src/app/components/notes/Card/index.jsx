import React from 'react';
import {Link} from 'react-router-dom';

import styles from '../../../../themes/style.scss';

class Card extends React.PureComponent {
    render() {
        return (
            <div className={`col-xs-1 col-md-4 col-xl-3 d-flex align-items-stretch ${styles.column}`} key={this.props.id}>
                <div className={`ui raised card ${styles.card}`}>
                    <div className={`content`}>
                        <i
                            onClick={() => this.props.deleteFunc(this.props.id)}
                            className={`right floated close icon ${styles["delete-button"]}`}
                        />
                        <div className={`header ${styles.header}`}>
                            {this.props.title}
                        </div>
                        <div className={`description ${styles.body}`}>
                            <p>
                                {this.props.body}
                            </p>
                        </div>
                    </div>
                    <div className={"extra content"}>
                        <span className="right floated pencil alternate">
                            {/*pass props to link*/}
                            <Link to={`/edit/${this.props.id}`}>
                                <i className={`pencil alternate icon ${styles["edit-button"]}`}/>
                            </Link>
                        </span>
                    </div>
                </div>
            </div>
        )
    }
}

export default Card;
