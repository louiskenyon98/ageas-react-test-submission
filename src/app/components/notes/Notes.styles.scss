.column {
  margin-bottom: 2rem;
  .card {
    max-height: 325px;
    margin: auto;
  }
  .card:hover {
    box-shadow: 0 10px 20px rgba(0,0,0,.09), 0 6px 6px rgba(0,0,0,.13);
  }
  .header {
    font-family: 'Thasadith', sans-serif !important;
  }
  .body {
    max-height: 170px;
    overflow: hidden;
    white-space: pre-wrap;
  }
}
.edit-button {
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  cursor: pointer
}
.edit-button:hover {
  color: cadetblue;
  transform: rotate(90deg);
}
.delete-button {
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  cursor: pointer
}
.delete-button:hover {
  color: crimson;
  transform: rotate(180deg);
}
