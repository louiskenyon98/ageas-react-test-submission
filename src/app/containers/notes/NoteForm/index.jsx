import React from 'react';
import {Field, reduxForm} from 'redux-form';
import {TextArea, TextInput} from '../../../components/common/reduxFormFields/index';

import styles from '../../../../themes/style.scss';
import {Link} from "react-router-dom";

export class NoteFormContainer extends React.Component {
    constructor(props) {
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(formValues) {
        this.props.onSubmit(formValues)
    };

    render() {
        return (
            <form onSubmit={this.props.handleSubmit(this.onSubmit)}
                  className={`note-editor-form ui form error ${styles.form}`}>
                <Field
                    name='title'
                    component={TextInput}
                    label='Enter Note Title...'
                />
                <Field
                    name='body'
                    component={TextArea}
                    label='Enter Note Body...'
                />
                <button className={`ui right floated button`}>Submit</button>
                <Link to="/">
                    <button className={`ui button`}>Cancel</button>
                </Link>
            </form>
        )
    }
}

export const validate = (formValues) => {
    const errors = {};
    if (!formValues.title) {
        errors.title = 'You must enter a title for your note'
    }
    if (!formValues.body) {
        errors.body = 'You must enter a body for your note'
    }
    return errors;
};
export default reduxForm({
    form: 'noteForm',
    validate,
    enableReinitialize: true
})(NoteFormContainer);
