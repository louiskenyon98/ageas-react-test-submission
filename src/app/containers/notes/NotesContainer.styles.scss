.noteCreator {
  width: 100%;
  margin-top: 10vh;
  h2 {
    text-align: center;
  }
}
.form {
  align-items: center !important;
  font-weight: 300;
  button {
    background-color: #3498db !important;
    color: #fff !important;
  }
  button:hover {
    box-shadow: 0 10px 20px rgba(0,0,0,.09), 0 6px 6px rgba(0,0,0,.13) !important;
  }
}
