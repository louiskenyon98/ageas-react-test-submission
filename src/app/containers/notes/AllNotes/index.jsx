import React from 'react';
import {connect} from 'react-redux';
import {getNotes, deleteNote} from '../../../actions';

import AllNotes from "../../../components/notes/AllNotes";
// import AllNotes from "../../components/AllNotes";

class AllNotesContainer extends React.Component {
    componentDidMount() {
        this.props.getNotes();
    }
    render() {
        return (
            <AllNotes
                notes={this.props.notes}
                deleteNote={this.props.deleteNote}
            />
        )
    }

}

const mapStateToProps = (state) => {
    return {
        notes: Object.values(state.notes)
    }
};

export default connect(mapStateToProps, {getNotes, deleteNote})(AllNotesContainer);
