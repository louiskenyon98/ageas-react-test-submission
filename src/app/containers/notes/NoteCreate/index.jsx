import React from 'react';
import {connect} from 'react-redux';
import {getNotes, postNote} from '../../../actions/note/notes.action';
import NoteFormContainer from '../NoteForm';
import styles from '../NotesContainer.styles.scss';


export class NoteCreateContainer extends React.Component {
    constructor(props) {
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(formValues) {
        this.props.postNote(formValues, this.props.getNotes);
    };

    render() {
        return (
            <div className={`col-xl-4 offset-xl-4 col-lg-6 offset-lg-3 col-sm-8 offset-sm-2 ${styles.noteCreator}`}>
                <h2>Create New Note</h2>
                <NoteFormContainer
                    onSubmit={this.onSubmit}
                />
            </div>

        )
    }
}

export default connect(
    null,
    {postNote, getNotes}
)(NoteCreateContainer);
