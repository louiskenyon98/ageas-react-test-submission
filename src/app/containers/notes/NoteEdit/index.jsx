import _ from 'lodash';
import React from 'react';
import {connect} from 'react-redux';
import {getNote, getNotes, patchNote} from '../../../actions/note/notes.action';
import NoteFormContainer from '../NoteForm';
import styles from '../../../../themes/style.scss';

export class NoteEditContainer extends React.Component {
    constructor(props) {
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
    }

    componentDidMount() {
        this.props.getNote(this.props.match.params.id);
    }

    onSubmit(formValues) {
        this.props.patchNote(formValues, this.props.getNotes)
    };

    render() {
        return (
            <div className={`col-xl-4 offset-xl-4 col-lg-6 offset-lg-3 col-sm-8 offset-sm-2 ${styles["note-creator"]}`}>
                <h2>Edit Note</h2>
                <NoteFormContainer
                    //Use Lodash to extract only the properties specified below from this.props.note.
                    initialValues={_.pick(this.props.note, 'id', 'title', 'body')}
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

NoteEditContainer.defaultProps = {
    note: {}
};


const mapStateToProps = (state) => {
    return {
        note: state.notes.edit
    }
};

export default connect(mapStateToProps, {getNote, getNotes, patchNote})(NoteEditContainer)
