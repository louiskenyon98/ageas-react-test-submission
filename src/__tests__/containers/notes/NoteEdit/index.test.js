import React from 'react';
import {shallow} from 'enzyme';
import {NoteEditContainer} from '../../../../app/containers/notes/NoteEdit';
import NoteFormContainer from '../../../../app/containers/notes/NoteForm/index';

describe('NoteEditContainer', () => {
    let props = {};
    let wrapper = () => shallow(<NoteEditContainer {...props}/>);
    beforeEach(() => {
        props = {
            getNote: jest.fn(),
            match: {
                isExact: true,
                params: {
                    id: 89
                },
                path: "/edit/:id",
                url: "/edit/89"
            },
            patchNote: jest.fn(),
            getNotes: jest.fn(),
            note: {
                id: 89,
                title: "noteTitle",
                body: "noteBody",
                createdAt: "2019-04-04T10:21:10.289Z",
                updatedAt: "2019-04-04T10:21:10.289Z"
            }
        }
    });
    describe('render', () => {
        it('should render NoteEditContainer with Redux Form as child', () => {
            expect(wrapper()).toMatchSnapshot();
        });
    });
    describe('functionality', () => {
        it('should call getNote on mount', () => {
            expect(props.getNote).not.toHaveBeenCalledWith();
            wrapper();
            expect(props.getNote).toHaveBeenCalled();
            expect(props.getNote).toHaveBeenCalledTimes(1);
        });
        it('should call patchNote onSubmit', () => {
            const formValues = {
                hello: 123
            };
            wrapper().find(NoteFormContainer).first().props().onSubmit(formValues);
            expect(props.patchNote).toHaveBeenCalledTimes(1);
            expect(props.patchNote).toHaveBeenCalledWith(formValues, expect.anything());
        })
    })
});
