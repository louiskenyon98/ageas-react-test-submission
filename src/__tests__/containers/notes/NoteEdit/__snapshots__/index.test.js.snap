// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteEditContainer render should render NoteEditContainer with Redux Form as child 1`] = `
<div
  className="col-xl-4 offset-xl-4 col-lg-6 offset-lg-3 col-sm-8 offset-sm-2 note-creator"
>
  <h2>
    Edit Note
  </h2>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={true}
    forceUnregisterOnUnmount={false}
    form="noteForm"
    getFormState={[Function]}
    initialValues={
      Object {
        "body": "noteBody",
        "id": 89,
        "title": "noteTitle",
      }
    }
    keepDirtyOnReinitialize={false}
    onSubmit={[Function]}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
    validate={[Function]}
  />
</div>
`;
